tutorials/haskell/gui-application/tutorial.md

```diff
@@ -0,0 +1,825 @@
+enterDot (Value x action) =
+  let f xs = if '.' `elem` xs then xs else '.' : xs
+  in case action of
+       Nothing -> Value (f x) Nothing
+       Just  a -> Value x (Just $ mapAction f a)
+
+-- | Change state as if specific char (digit) is entered.
+enterDigit :: Char -> Value -> Value
+enterDigit ch (Value x action) =
+  case action of
+    Nothing -> Value (ch:x) Nothing
+    Just  a -> Value x (Just $ mapAction (ch:) a)
+
+-- | Change state as if last character of current argument is removed.
+backspace :: Value -> Value
+backspace (Value x action) =
+  case action of
+    Nothing -> Value (drop 1 x) Nothing
+    Just  a -> Value x (Just $ mapAction (drop 1) a)
+
+-- | Apply given operator to current state. If some action is already fully
+-- constructed, evaluate it first.
+operator :: (String -> Action) -> Value -> Value
+operator op value =
+  let (Value x action) = equals value
+  in Value x $ Just $
+    case action of
+      Nothing -> op ""
+      Just  a -> op (getSndArg a)
+
+-- | Change state as if current argument is removed.
+clearEntry :: Value -> Value
+clearEntry (Value x action) =
+  case action of
+    Nothing -> Value "" Nothing
+    Just  a ->
+      if null (getSndArg a)
+        then Value "" Nothing
+        else Value x (Just $ mapAction (const "") a)
+
+-- | Change state returning it to the default value.
+clearAll :: Value -> Value
+clearAll = const (Value "" Nothing)
+
+-- | Evaluate current calculator's state putting result in place of first
+-- argument.
+equals :: Value -> Value
+equals (Value x action) =
+  case action of
+    Nothing -> Value x Nothing
+    Just  a ->
+      if null (getSndArg a)
+        then Value x action
+        else Value result Nothing
+          where
+            g  :: String -> Double
+            g ""       = 0
+            g ('.':xs) = g ('0':'.':xs)
+            g xs       = read (reverse xs)
+            x' = g x
+            y' = g (getSndArg a)
+            result = reverse . show $
+              case a of
+                Addition       _ -> x' + y'
+                Subtraction    _ -> x' - y'
+                Multiplication _ -> x' * y'
+                Division       _ -> x' / y'
+```
+
+The calculator is not perfect, but good enough for our purposes. Compile,
+run it and see how it works. Implementation of the rest of functionality is
+left as an exercise for the reader.
+
+## Using Glade to design forms
+
+You probably have noticed that manual creation of forms introduces quite a
+bit of boilerplate in our code. This can become even worse as your forms get
+more complex. Because of this, I think it's time to try our hand on a modern
+UI designer for GTK+ called [Glade](https://glade.gnome.org/).
+
+Glade is straightforward to install and use. Open the application, you will
+see panels with various widgets: top-level objects (such as window),
+containers, and controls (such as buttons).
+
+Short plan how to re-create our calculator form with Glade:
```

Comment:
@mrkkrp Can you rephrase this? Maybe "Here's the short plan to re-create our calculator form with Glade".